android/sdk/src/main/java/app/organicmaps/sdk/bookmarks/data/MapObject.java

```diff
@@ -107,32 +107,32 @@ public MapObject(@NonNull FeatureId featureId, @MapObjectType int mapObjectType,
     mWikiArticle = wikiArticle;
     mRoadWarningMarkType = RoadWarningMarkType.values()[roadWarningType];
     if (rawTypes != null)
-      mRawTypes = new ArrayList<>(Arrays.asList(rawTypes));
+      mRawTypes = Arrays.asList(rawTypes);
   }
 
+  // Also called from Bookmark constructor.
   protected MapObject(@MapObjectType int type, Parcel source)
   {
-    this(Objects.requireNonNull(
-             ParcelCompat.readParcelable(source, FeatureId.class.getClassLoader(), FeatureId.class)), // FeatureId
-         type, // MapObjectType
-         source.readString(), // Title
-         source.readString(), // SecondaryTitle
-         source.readString(), // Subtitle
-         source.readString(), // Address
-         source.readDouble(), // Lat
-         source.readDouble(), // Lon
-         ParcelCompat.readParcelable(source, Metadata.class.getClassLoader(), Metadata.class),
-         source.readString(), // ApiId;
-         ParcelCompat.readParcelable(source, RoutePointInfo.class.getClassLoader(),
-                                     RoutePointInfo.class), // RoutePointInfo
-         source.readInt(), // mOpeningMode
-         Objects.requireNonNull(
-             ParcelCompat.readParcelable(source, Popularity.class.getClassLoader(), Popularity.class)),
-         Objects.requireNonNull(source.readString()), source.readInt(),
-         null // mRawTypes
-    );
-
-    mRawTypes = readRawTypes(source);
+    // Type has already been read in readFromParcel method.
+    mMapObjectType = type;
+    // Reading order must be the same as writing order in writeToParcel.
+    mFeatureId =
+        Objects.requireNonNull(ParcelCompat.readParcelable(source, FeatureId.class.getClassLoader(), FeatureId.class));
+    mTitle = source.readString();
+    mSecondaryTitle = source.readString();
+    mSubtitle = source.readString();
+    mAddress = source.readString();
+    mLat = source.readDouble();
+    mLon = source.readDouble();
+    mMetadata = ParcelCompat.readParcelable(source, Metadata.class.getClassLoader(), Metadata.class);
+    mApiId = source.readString();
+    mRoutePointInfo = ParcelCompat.readParcelable(source, RoutePointInfo.class.getClassLoader(), RoutePointInfo.class);
+    mOpeningMode = source.readInt();
+    // mPopularity =
+    //     ParcelCompat.readParcelable(source, Popularity.class.getClassLoader(), Popularity.class);
+    mWikiArticle = source.readString();
+    mRoadWarningMarkType = RoadWarningMarkType.values()[source.readInt()];
+    source.readStringList(mRawTypes);
```

Comment:
Why not to keep original structure with `this(...)` call? Where is the difference?